### PR TITLE
Pardiso

### PR DESCRIPTION
**Pull request by vpuri3**
todo

- [x] finish boilerplate
- [ ] test on some realistic sciml usecase
- [x] test on pardiso.jl/examples

close https://github.com/SciML/LinearSolve.jl/issues/5

### CODE REVIEW

**Comment by vpuri3** [comment]
@ChrisRackauckas i got some boilerplate code with a little functionality. can you point me to usecases in sciml ecosystem to test against so i can figure what more to add if needed? else i can test this with the examples in https://github.com/JuliaSparse/Pardiso.jl/tree/master/examples

File: Project.toml [hunk 1]
```diff
@@ -9,8 +9,10 @@ IterativeSolvers = "42fd0dbc-a981-5370-80f2-aaf504508153"
 Krylov = "ba0b0d4f-ebba-5204-a429-3ac8c609bfb7"
 KrylovKit = "0b1a1467-8014-51b9-945f-bf0ae24f4b77"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Pardiso = "46dd5b70-b6fb-5a00-ae2d-e8fea33afaf2"
```
Comment (ChrisRackauckas) [inline Project.toml:12]: remove from the dependencies

File: src/wrappers.jl [hunk 2]
```diff
@@ -297,3 +297,71 @@ function SciMLBase.solve(cache::LinearCache, alg::IterativeSolversJL; kwargs...)
...
+    """
+    figure out whatever phase is. should set_phase call be in init_cacheval?
+    can we use phase to store factorization in cache?
+    """
+    Pardiso.set_phase!(cacheval, alg.solve_phase)
+    Pardiso.solve!(cacheval, u, A, b)
+    Pardiso.set_phase!(cacheval, alg.release_phase) # is this necessary?
+
+    return cache.u
+end
```
Comment (ChrisRackauckas) [inline src/wrappers.jl:367]: this should be in a `pardiso.jl` that is included only when `@requires`'d

File: src/LinearSolve.jl [hunk 3]
```diff
@@ -38,14 +40,16 @@ function __init__()
...
 
 export LUFactorization, SVDFactorization, QRFactorization, GenericFactorization,
        RFLUFactorizaation
-export KrylovJL, KrylovJL_CG, KrylovJL_GMRES, KrylovJL_BICGSTAB,
-       KrylovJL_MINRES,
+export KrylovJL, KrylovJL_CG, KrylovJL_GMRES, KrylovJL_BICGSTAB, KrylovJL_MINRES,
        IterativeSolversJL, IterativeSolversJL_CG, IterativeSolversJL_GMRES,
-       IterativeSolversJL_BICGSTAB, IterativeSolversJL_MINRES
+       IterativeSolversJL_BICGSTAB, IterativeSolversJL_MINRES,
+       PardisoJL
```
Comment (ChrisRackauckas) [inline src/LinearSolve.jl:52]: the export will need to be in the requires statement

File: src/wrappers.jl (hunk 2 above)
Comment (vpuri3) [reply src/wrappers.jl]: done

File: src/LinearSolve.jl (hunk 3 above)
Comment (vpuri3) [reply src/LinearSolve.jl]: done

File: src/LinearSolve.jl [hunk 4]
```diff
@@ -38,12 +39,13 @@ function __init__()
   else
     IS_OPENBLAS[] = occursin("openblas", BLAS.get_config().loaded_libs[1].libname)
   end
+
+  @require Pardiso="46dd5b70-b6fb-5a00-ae2d-e8fea33afaf2" PardisoJL
```
Comment (ChrisRackauckas) [inline src/LinearSolve.jl:43]: needs to be an include. See the DiffEqBase requires handling as an example.

Comment (vpuri3) [reply src/LinearSolve.jl]: missed it sorry. fixed now

File: src/pardiso.jl [hunk 5]
```diff
@@ -0,0 +1,74 @@
+
+## Paradiso
+
+import Pardiso
+
+export PardisoJL, PardisoJLFactorize, PardisoJLIterate
```
Comment (ChrisRackauckas) [inline src/pardiso.jl:6]: doesn't pardiso have a mixed factorized and iterate thing?

Comment (vpuri3) [reply src/pardiso.jl:6]: yeah - ANALYSIS_NUM_FACT_SOLVE_REFINE

Comment (vpuri3) [reply src/pardiso.jl:6]: doing that too

**Comment by vpuri3** [comment]
@Kristoffer Carlsson , we're writing a wrapper for PardisoJL for DiffEq interface. Would like for you to have a look at this. I was also wondering how one could control `abstol`, and `reltol`. Would it be possible by supplying appropriate `iparm/dparms`??

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
This looks good to merge. Of course, more can be tweaked in further PRs as better defaults are found.

**Pull request merged**